src/python/pants/backend/python/goals/export.py

```diff
@@ -334,12 +356,155 @@ class MaybeExportResult:
+                            os.rename(os.path.join(codegen_dir, item), os.path.join(site_packages_dir, item))
+                        """
+                    ).encode(),
+                )
+            ]
+        ),
+    )
+
+    return _ExportPythonCodegenSetup(codegen_setup_script_digest)
+
+
+@rule
+async def export_python_codegen(
+    request: _ExportPythonCodegenRequest, python_setup: PythonSetup, all_targets: AllTargets
+) -> _ExportPythonCodegenResult:
+    non_python_sources_in_python_resolve = [
+        tgt.get(SourcesField)
+        for tgt in all_targets
+        if tgt.has_field(PythonResolveField)
+        and tgt[PythonResolveField].normalized_value(python_setup) == request.resolve
+        and tgt.has_field(SourcesField)
+        and not tgt.has_field(PythonSourceField)
+    ]
+
+    if not non_python_sources_in_python_resolve:
+        return _ExportPythonCodegenResult(EMPTY_DIGEST)
+
+    hydrated_non_python_sources = await MultiGet(
+        Get(
+            HydratedSources,
+            HydrateSourcesRequest(
+                sources,
+                for_sources_types=(PythonSourceField,),
+                enable_codegen=True,
+            ),
+        )
+        for sources in non_python_sources_in_python_resolve
+    )
+
+    merged_snapshot = await Get(
+        Snapshot,
+        MergeDigests(
+            hydrated_sources.snapshot.digest for hydrated_sources in hydrated_non_python_sources
+        ),
+    )
+
+    stripped_source_files = await Get(StrippedSourceFiles, SourceFiles(merged_snapshot, ()))
+
+    return _ExportPythonCodegenResult(stripped_source_files.snapshot.digest)
+
+
+# Generate codegen Python sources and add them to the virtualenv to be exported.
+async def add_codegen_to_export_result(
+    resolve: str, export_result: ExportResult, codegen_setup: _ExportPythonCodegenSetup
+) -> ExportResult:
+    # Generate Python sources from codegen targets in this resolve.
+    codegen_result = await Get(
+        _ExportPythonCodegenResult, _ExportPythonCodegenRequest(resolve=resolve)
+    )
+    if codegen_result.digest == EMPTY_DIGEST:
+        return export_result
+
+    codegen_digest = await Get(
+        Digest, AddPrefix(codegen_result.digest, _ExportPythonCodegenSetup.PKG_DIR)
+    )
+
+    export_digest_with_codegen = await Get(
+        Digest,
+        MergeDigests([export_result.digest, codegen_digest, codegen_setup.setup_script_digest]),
+    )
+
+    codegen_post_processing_cmds = (
+        PostProcessingCommand(
+            [
+                os.path.join("{digest_root}", "bin", "python"),
+                os.path.join(
+                    "{digest_root}",
+                    _ExportPythonCodegenSetup.PKG_DIR,
+                    _ExportPythonCodegenSetup.SCRIPT_NAME,
+                ),
+                os.path.join("{digest_root}", _ExportPythonCodegenSetup.PKG_DIR),
+            ]
+        ),
+        PostProcessingCommand(
+            [
+                "rm",
+                "-rf",
```

Comment:
This is a little terrifying to me. But isn't PKG_DIR supposed to be empty now, apart from the script, so we could rm the script and rmdir the empty dir ?

Reply:
Good point. Direct `rm` and `rmdir` will work just fine. Will switch to that.